Rephrasing MPD registry.md (#945)
plone/training: docs/mastering-plone/registry.md

Most settings in `/plone_control_panel` reside in the registry. You can modify them directly through its UI.

Open http://localhost:8080/Plone/portal_registry and filter for `displayed_types`.
Modify the content types shown in the navigation and site map directly.
These values match those in http://localhost:8080/Plone/@@navigation-controlpanel, where the form is customized for better usability.

```{note}
This UI for the registry is not yet available in the frontend.
```

## Registry records

In {doc}`volto_frontpage` you already added a criterion usable for listing blocks in {file}`profiles/default/registry/querystring.xml`.
This setting is stored in the registry.

Examine the existing values in the registry.

Go to http://localhost:3000/controlpanel/navigation and add `talk` to the field {guilabel}`Displayed content types`.
Talks in the root will now show up in the navigation.
This setting is stored in the registry record `plone.displayed_types`.

## Accessing and modifying records in the registry

> In Python you can access the registry record with the key `plone.displayed_types` via {py:mod}`plone.api`.

Author: Thanks for the hint. I add a link to the plone.api documentation.

> In Python you can access the registry record with the key `plone.displayed_types` via {py:mod}`plone.api`.

Author: done in https://github.com/plone/training/pull/945/commits/ef2016a04353f4839a60bb399b132c74c20c6ed8

## Introduction

Do you remember the fields `audience` and `type_of_talk` in the talk content type?
The schema previously hard-coded several options for selection.

> Next, you will add a field to assign talks to a room. 

Reviewer: ```suggestion
Next, you will add a field to assign talks to a room.
```

Most settings in `/plone_control_panel` reside in the registry. You can modify them directly through its UI.

Open http://localhost:8080/Plone/portal_registry and filter for `displayed_types`.
Modify the content types shown in the navigation and site map directly.
These values match those in http://localhost:8080/Plone/@@navigation-controlpanel, where the form is customized for better usability.

```{note}
This UI for the registry is not yet available in the frontend.
```

## Registry records

In {doc}`volto_frontpage` you already added a criterion usable for listing blocks in {file}`profiles/default/registry/querystring.xml`.
This setting is stored in the registry.

Examine the existing values in the registry.

Go to http://localhost:3000/controlpanel/navigation and add `talk` to the field {guilabel}`Displayed content types`.
Talks in the root will now show up in the navigation.
This setting is stored in the registry record `plone.displayed_types`.

## Accessing and modifying records in the registry

> In Python you can access the registry record with the key `plone.displayed_types` via {py:mod}`plone.api`.portal.

Author: I call it by intent `plone.api`.portal, cause `plone.api` is THE module.

> In Python you can access the registry record with the key `plone.displayed_types` via {py:mod}`plone.api`.portal.

Reviewer 2: plone.api is a package which contains the plone.api.portal module. We can link to either one here since the specific functions are shown below. But we should label it with the same name that we link to.

> In Python you can access the registry record with the key `plone.displayed_types` via {py:mod}`plone.api`.portal.

Reviewer: We also need to use correct syntax so that it both has a link (it currently does not), and links to the correct location. The current rendering is wrong, as shown in the screenshot.

![Screenshot 2025-04-18 at 1 28 28 PM](https://github.com/user-attachments/assets/9433e6b7-73d7-4307-bcda-cccdc0dce660)

Alternatively, be explicit.

```
In Python you can access the registry record with the key `plone.displayed_types` via the {py:mod}`~plone.api.portal` module in `plone.api`.
```

Or you can use an inline literal for the whole thing.

Whatever you choose, check its rendering at https://plone-training--945.org.readthedocs.build/mastering-plone/registry.html#accessing-and-modifying-records-in-the-registry.

## Introduction

Do you remember the fields `audience` and `type_of_talk` in the talk content type?
The schema previously hard-coded several options for selection.

Next, you will add a field to assign talks to a room.
The room names change each year for the conference, so site administrators need to edit them.

Additionally, admins should be able to edit the options for `audience` and `type_of_talk`, making it possible to add options like _Lightning Talks_!

By combining the registry, a control panel and vocabularies you can make rooms configurable options.

To achieve this you first need to get to know the registry.

## The registry

The registry stores and retrieves values in records.
Each record consists of the actual value, along with a field that describes the record in more detail.
You can interact with the registry using Python dictionary-style operations to get and set values.

Since Plone 5 the registry stores all global settings.
Plone provides the registry through [plone.registry](https://pypi.org/project/plone.registry) and offers a user interface for interaction via [plone.app.registry](https://pypi.org/project/plone.app.registry).

> Most settings in `/plone_control_panel` reside in the registry.

Reviewer 2: Let's use the name that appears in the UI. I don't recall that there was ever something called `/plone_control_panel`.

```suggestion
Most settings in Site Setup reside in the registry.
```

You can modify them directly through its UI.

Open http://localhost:8080/Plone/portal_registry and filter for `displayed_types`.
Modify the content types shown in the navigation and site map directly.
These values match those in http://localhost:8080/Plone/@@navigation-controlpanel, where the form is customized for better usability.

```{note}
This UI for the registry is not yet available in the frontend.
```

## Registry records

In {doc}`volto_frontpage` you already added a criterion usable for listing blocks in {file}`profiles/default/registry/querystring.xml`.
This setting is stored in the registry.

Examine the existing values in the registry.

Go to http://localhost:3000/controlpanel/navigation and add `talk` to the field {guilabel}`Displayed content types`.
Talks in the root will now show up in the navigation.
This setting is stored in the registry record `plone.displayed_types`.

## Accessing and modifying records in the registry

In Python you can access the registry record with the key `plone.displayed_types` via {py:mod}`plone.api`.portal.

> It holds convenience methods to `get` and `set` a record:

Reviewer 2: ```suggestion
It holds convenience functions to `get` and `set` a record:
```

A method is a function that invoked on a class instance.

## Introduction

Do you remember the fields `audience` and `type_of_talk` in the talk content type?
The schema previously hard-coded several options for selection.

Next, you will add a field to assign talks to a room.
The room names change each year for the conference, so site administrators need to edit them.

Additionally, admins should be able to edit the options for `audience` and `type_of_talk`, making it possible to add options like _Lightning Talks_!

By combining the registry, a control panel and vocabularies you can make rooms configurable options.

To achieve this you first need to get to know the registry.

## The registry

The registry stores and retrieves values in records.
Each record consists of the actual value, along with a field that describes the record in more detail.
You can interact with the registry using Python dictionary-style operations to get and set values.

Since Plone 5 the registry stores all global settings.
Plone provides the registry through [plone.registry](https://pypi.org/project/plone.registry) and offers a user interface for interaction via [plone.app.registry](https://pypi.org/project/plone.app.registry).

> Most settings in Site Setup reside in the registry.

Reviewer: ```suggestion
Most settings in {guilabel}`Site Setup` reside in the registry.
```

You can modify them directly through its UI.

Open http://localhost:8080/Plone/portal_registry and filter for `displayed_types`.
Modify the content types shown in the navigation and site map directly.
These values match those in http://localhost:8080/Plone/@@navigation-controlpanel, where the form is customized for better usability.

```{note}
This UI for the registry is not yet available in the frontend.
```

## Registry records

In {doc}`volto_frontpage` you already added a criterion usable for listing blocks in {file}`profiles/default/registry/querystring.xml`.
This setting is stored in the registry.

Examine the existing values in the registry.

Go to http://localhost:3000/controlpanel/navigation and add `talk` to the field {guilabel}`Displayed content types`.
Talks in the root will now show up in the navigation.
This setting is stored in the registry record `plone.displayed_types`.

## Accessing and modifying records in the registry

In Python you can access the registry record with the key `plone.displayed_types` via `plone.api.portal`.

> It holds convenience funtions to `get` and `set` a record:

Reviewer: These are verbs, not inline literals. 

```suggestion
It holds convenience functions to get and set a record:
```

api.portal.set_registry_record('plone.smtp_host', 'my.mail.server')
```

> For more information see {py:mod}`plone.api`.portal documentation {ref}`plone6docs:portal-get-registry-record-example`.

Author: I call it by intent `plone.api`.portal, cause `plone.api` is THE module.

> For more information see {py:mod}`plone.api`.portal documentation {ref}`plone6docs:portal-get-registry-record-example`.

Reviewer 2: I don't understand your point, @ksuess.

]
```

> If the name "Lightning-Talk" needs to be updated to 'Short talks', the talks categorized as lightning talks will still display correctly.

Author: I also like it better

> This is because the value stored in the talks is the token "lightning-talk," which remains unchanged.

Author: Well, this Vale warning that punctuation should go inside quotes was and is confusing. 
But now, finally, I understand, that quotes are for phrases, not for single word values.
Instead i choose to go with inline literals (with backticks) from now on.

> This is because the value stored in the talks is the token "lightning-talk," which remains unchanged.

Reviewer: Vale expects this pattern for commas inside quote marks.

"Vale is nitpicky and may not be correct all the time," said Steve.

The comma is used here to show that there is more to the sentence.

Whereas there are things such as "air quotes", where the comma goes after the quoted term.

See https://www.microsoft.com/en-us/microsoft-365-life-hacks/writing/quotation-mark-rules for more confusing English punctuation.

You can use `ploneconf.rooms` instead of `ploneconf.site.controlpanel.controlpanel.IPloneconfSettings.rooms`.
```

After reinstalling the package to apply the registry changes, you can access and modify these registry records as described before.

> Either use http://localhost:8080/Plone/portal_registry or `Python`:

Author: TIL

## Add a custom control panel

Now you will add a custom control panel to edit all settings related to the package with a user-friendly interface.

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Author: Indeed I stripped in 2023 everything in this training concerning only ClassicUI, but not Volto. So this should read `To register a control panel for the frontend, you need a bit of boiler-plate:`.

Question for correct english: with or without commma? 

Not that it matters, but…
In german I would go with comma, cause the first part is a `Nebensatz`.

_Der Satz "Um ein Kontroll-Panel zu registrieren, brauchst du einen Hammer" ist korrekt mit Komma geschrieben.
In diesem Fall wird das Komma gesetzt, weil ein Nebensatz (Infinitivgruppe "Um ein Kontroll-Panel zu registrieren") dem Hauptsatz ("brauchst du einen Hammer") vorangestellt ist. Nach der deutschen Rechtschreibung wird bei erweiterten Infinitivgruppen mit "um zu", "ohne zu" oder "statt zu" in der Regel ein Komma gesetzt.
Die korrekte Schreibweise ist also:
"Um ein Kontroll-Panel zu registrieren, brauchst du einen Hammer"_
Claude 3.7 Sonnet

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Author: > Indeed I stripped in 2023 everything in this training concerning only ClassicUI, but not Volto.

I add a note, that, if accessing the control panel in ClassicUI is needed, instructions can be found in training version `Volto + ClassicUI` https://2022.training.plone.org/mastering-plone/registry.html#add-a-custom-control-panel

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Reviewer: This is correct:

`To register a control panel, you need a bit of boilerplate code.`

Also it is `Classic UI`, with a space, unless it is an inline literal.

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Reviewer: In English, the comma represents a breath when speaking or a list of items. I think the same applies in German.

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Author: I prefer `ClassicUI`. 
Claude confirms with:

_# ClassicUI vs. Classic UI: Why One Word Matters for Documentation Searching

When looking for technical documentation, the search terms you use can significantly impact your results. I've found that searching for "ClassicUI" (as one word) instead of "Classic UI" (as two words) makes a noticeable difference in finding relevant information.

## Why I prefer "ClassicUI" as one word:

1. **More precise search results** - Search engines treat "ClassicUI" as a specific technical term rather than two common words ("Classic" and "UI") that might appear separately in unrelated contexts.

2. **Matches actual code syntax** - Many programming libraries and frameworks use camelCase or PascalCase naming conventions, making "ClassicUI" more likely to match the actual code implementation.

3. **Reduces noise in search results** - "Classic UI" might return general articles about traditional user interfaces, while "ClassicUI" targets the specific component or framework you're looking for.

4. **Aligns with documentation standards** - Technical documentation often follows the same naming conventions as the code itself, making single-word searches more effective.

5. **Helps find code repositories** - When searching GitHub or other code repositories, the one-word format better matches variable names, class names, and repository names.

Next time you're struggling to find documentation on a specific UI component or framework, try searching for the concatenated version of the term. It might just help you find exactly what you're looking for faster!_

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Reviewer: That's not consistent with every other instance throughout documentation. "ClassicUI" expanded is "ClassicUserInterface", which is just plain wrong. There is no way that we're going to rewrite all documentation because Claude says to do it.

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Author: You nailed it.
The 2022 training includes instructions how to add the control panel, created according https://training.plone.org/mastering-plone/registry.html, to the classic user interface https://classic.demo.plone.org/@@overview-controlpanel

> To register a control panel for the frontend and Plone Classic you need a bit of boiler-plate:

Author: To be consistent: "Classic UI", with a space

    group = "Products"
```

> If you want to use this control panel in ClassicUI as well, see https://2022.training.plone.org/mastering-plone/registry.html#add-a-custom-control-panel, which also handles the ClassicUI version.

Reviewer: ```suggestion
If you want to use this control panel in Classic UI as well, see https://2022.training.plone.org/mastering-plone/registry.html#add-a-custom-control-panel, which also handles the Classic UI version.
```

## Vocabularies

Now the custom settings are stored in the registry and can be modified conveniently by site administrators.
These options still need to be used in talks.

To achieve this, turn them into vocabularies.

Vocabularies are often used for selection fields.
They have many benefits:

- They enable you to separate the select option values from the content type schema.

>   Users can edit vocabularies through the UI.  
> - Developers can set vocabularies dynamically.  

Reviewer: ```suggestion
  Users can edit vocabularies through the UI.
- Developers can set vocabularies dynamically.
```

    component="ploneconf.site.vocabularies.talk.RoomsVocabularyFactory" />
```

> From now on you can use these vocabulary by referring to their name, for example `ploneconf.rooms`.

Reviewer: Editorial comment: I always die a little bit inside when I see `e.g.` used without a trailing comma. That or any of its mutations that are not `, e.g.,` between the leading sentence and its following example.

```suggestion
From now on you can use these vocabulary by referring to their name, for example, `ploneconf.rooms`.
```

## Summary

- You successfully combined the registry, a control panel, and vocabularies to enable site administrators to manage field options.

> - It seems like a lot, but you will certainly use dynamic vocabularies, control panels and the registry in many of your future Plone projects in one way or another.

Reviewer: Oxford comma.

```suggestion
- It seems like a lot, but you will certainly use dynamic vocabularies, control panels, and the registry in many of your future Plone projects in one way or another.
```
